flamingo/scripts/interpolate_X_Ray_redshift.py

```diff
@@ -0,0 +1,372 @@
+            print('Or limit your particle data set to be within the interpolation range')
+            print('\n')
+            print('Interpolation will be capped, all particles outside of the interpolation range will be set to the edges of that range')
+            joint_mask = data_n > -100 #Everything
+        else:
+            emissivities[~joint_mask] = fill_value
+
+
+    #get individual mass fraction
+    mass_fraction = df_elements.to_numpy()
+
+    # # From Cooling tables, integrate into X-Ray tables in a future version
+    # max_mass_fractions = np.array([7.5597578e-01, 2.5954419e-01, 7.1018077e-03, 2.0809614e-03,
+    #                                    1.7216533e-02, 3.7735226e-03, 2.1262325e-03, 1.9969980e-03,
+    #                                     3.8801527e-03])
+
+    # # At the moment the tables only go up to 0.5 * solar metallicity for He
+    # # Enforce this for all metals to have consistency 
+    # clip_mass_fractions = max_mass_fractions
+    # # Reset such to mass fractions of Hydrogen = 1
+    # clip_mass_fractions /= clip_mass_fractions[0]
+    # mass_fraction = np.clip(mass_fraction, a_min = 0, a_max = clip_mass_fractions)
+    
+
+    #Find density offsets
+    idx_n = find_idx(data_n[joint_mask], interp.density_bins)
+    dx_n = find_dx(data_n[joint_mask], interp.density_bins, idx_n[:, 0].astype(int))
+
+    #Find temperature offsets
+    idx_T = find_idx(data_T[joint_mask], interp.temperature_bins)
+    dx_T = find_dx(data_T[joint_mask], interp.temperature_bins, idx_T[:, 0].astype(int))
+
+    #Find element offsets
+    #mass of ['hydrogen', 'helium', 'carbon', 'nitrogen', 'oxygen', 'neon', 'magnesium', 'silicon', 'iron']
+    element_masses = [1, 4.0026, 12.0107, 14.0067, 15.999, 20.1797, 24.305, 28.0855, 55.845]
+
+    #Calculate the abundance wrt to solar
+    # abundances = mass_fraction / np.array(element_masses)
+    abundances = (mass_fraction / np.expand_dims(mass_fraction[:, 0], axis = 1)) / np.array(element_masses)
+
+    # Clip to solar metallicity
+    # abundances = np.clip(abundances, a_min = 0, a_max = 10**interp.solar_metallicity)
+
+    #Calculate abundance offsets using solar mass fractions
+    # This should be less susceptible to changes in the hydrogen mass fraction
+    solar_mass_fractions = [7.0030355e-01, 2.5954419e-01, 7.1018077e-03, 2.0809614e-03,
+                             1.7216533e-02, 3.7735226e-03, 2.1262325e-03, 1.9969980e-03,
+                             3.8801527e-03]
+    abundance_to_solar = 1 - mass_fraction / solar_mass_fractions
+
+    # Calculate abundance offsets using solar abundances
+    # abundance_to_solar = 1 - abundances / 10**interp.solar_metallicity
+
+    abundance_to_solar = np.c_[abundance_to_solar[:, :-1], abundance_to_solar[:, -2], abundance_to_solar[:, -2], abundance_to_solar[:, -1]] #Add columns for Calcium and Sulphur and add Iron at the end
+
+    #Find helium offsets
+    idx_he = find_idx_he(np.log10(abundances[:, 1]), interp.He_bins)

```

Comment:
@jemme07 There is an inconsistency here: while `idx_n` and `idx_T` are determined using the mask `joint_mask`, `idx_he` is not. As a result, `idx_n/idx_T` and `idx_he` do not have the same length. This does not give any errors because you use the shorter length (that of `idx_n`) to determine the loop count in `get_table_interp()`, but it does mean that there is a mismatch between the elements in `idx_n/idx_T` and `idx_he`. In other words: you are probably using the wrong `idx_he` for most of the elements during the interpolation.

Reply:
The interpolation code was written by Joey Braspenning. I'll contact him to take a look. Well spotted!